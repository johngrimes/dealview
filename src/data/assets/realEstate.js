// @flow

import moment from 'moment'

import { getDatabase, save as saveToDatabase, load as loadFromDatabase, loadAll as loadAllFromDatabase } from '../database.js'
import { AddressEmpty } from '../commonTypes.js'
import { ValuationsEmpty } from '../../components/forms/ValuationsInput.js'
import type { Values } from '../database.js'
import type { AddressValues } from '../commonTypes.js'
import type { Valuation, Valuations } from '../../components/forms/ValuationsInput.js'
import type { AssetValues } from './asset.js'

export type RealEstateValues = {
  id: string,
  name: string,
  address: AddressValues,
  notes: string,
  valuations: Valuations
}

export const RealEstateEmpty = {
  id: '',
  name: '',
  address: AddressEmpty,
  notes: '',
  valuations: ValuationsEmpty
}

export const save = (values: Values): Promise<string> => {
  return saveToDatabase('Asset.RealEstate', values)
}

export const load = (id: string): Promise<RealEstateValues> => {
  return loadFromDatabase('Asset.RealEstate', id)
}

export const loadAll = (): Promise<RealEstateValues[]> => {
  return loadAllFromDatabase('Asset.RealEstate')
}

export const assetSubscriber = (eventType: string, content: RealEstateValues): void => {
  getAssetRecord(content.id).then(existingAsset => {
    const lastValuation = getLastValuation(content)
    const asset: AssetValues = {
      type: 'RealEstate',
      instanceId: content.id,
      name: content.name
    }
    if (lastValuation) { asset.lastValuation = lastValuation.amount }
    if (existingAsset) { asset.id = existingAsset.id }
    saveToDatabase('Asset', asset).catch(error => { console.error(error) })
  })
}

export const getLastValuation = (realEstate: RealEstateValues): Valuation|false => {
  if (realEstate.valuations.length === 0) {
    return false
  }
  const sortedValuations = realEstate.valuations.slice().sort(compareValuationsByDate)
  const idxFuture = sortedValuations.findIndex(v => { moment(v.date, 'YYYY-MM-DD').valueOf() > moment().valueOf() })
  const slicedValuations = sortedValuations.slice(0, idxFuture)
  return slicedValuations[slicedValuations.length - 1]
}

const compareValuationsByDate = (a: Valuation, b: Valuation): number => {
  const [milliA, milliB] = [a, b].map(v => { return v.date ? moment(v.date).valueOf() : 0 })
  return milliA - milliB
}

const getAssetRecord = (realEstateId: string): Promise<AssetValues|false> => {
  return new Promise((resolve, reject) => {
    getDatabase().then(db => {
      const transaction = db.transaction(['Asset'], 'readonly')
      const store = transaction.objectStore('Asset')
      const index = store.index('type, instanceId')
      const request = index.get(['RealEstate', realEstateId])
      request.onsuccess = event => { resolve(event.target.result) }
      request.onerror = () => { resolve(false) }
    })
  })
}
