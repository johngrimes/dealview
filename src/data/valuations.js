import moment from 'moment'

import { DateStorageFormat } from './commonTypes.js'

export const ValuationDefault = {
  date: moment().format(DateStorageFormat),
  note: '',
  type: 'none',
}

export const PurchaseDefault = {
  note: 'Purchase amount',
  type: 'purchase',
}

export const SaleDefault = {
  note: 'Sale amount',
  type: 'sale',
}

export const purchaseFilter = v => v.type === 'purchase'
export const purchaseNegFilter = v => v.type !== 'purchase'
export const saleFilter = v => v.type === 'sale'
export const saleNegFilter = v => v.type !== 'sale'

export const compareValuationsByDate = (a, b) => {
  const [ milliA, milliB ] = [ a, b ].map(v => {
    return v.date ? moment(v.date, DateStorageFormat).valueOf() : 0
  })
  return milliA - milliB
}

export const updateValuationsWithPurchaseOrSale = (valuations, purchaseOrSale, type) => {
  const found = valuations.find(type === 'purchase' ? purchaseFilter : saleFilter)
  const updatedPurchaseOrSale = {
    ...(type === 'purchase' ? PurchaseDefault : SaleDefault),
    ...found,
    ...purchaseOrSale,
  }
  return valuations.filter(type === 'purchase' ? purchaseNegFilter : saleNegFilter).concat([updatedPurchaseOrSale])
}

export const updateValuationsWithPurchase = (valuations, purchase) =>
  updateValuationsWithPurchaseOrSale(valuations, purchase, 'purchase')

export const updateValuationsWithSale = (valuations, sale) =>
  updateValuationsWithPurchaseOrSale(valuations, sale, 'sale')
