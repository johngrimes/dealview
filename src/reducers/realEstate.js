// @flow

import { createObjectsReducer, InitialObjectStoreState } from 'reducers/objects'
import type { ObjectStoreState } from 'reducers/objects'
import type { RealEstateMap } from 'types/assets/realEstate'
import type { RealEstateAction } from 'actions/realEstate'

export type RealEstateState = ObjectStoreState & { +objects: RealEstateMap }

export const InitialRealEstateState: RealEstateState = InitialObjectStoreState

const RealEstateReducer: (RealEstateState, RealEstateAction) => RealEstateState =
  createObjectsReducer('REAL_ESTATE', 'REAL_ESTATE')

export default RealEstateReducer
