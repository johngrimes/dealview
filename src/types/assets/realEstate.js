// @flow

import { AddressDefaults } from 'types/commonTypes'
import type { Address } from 'types/commonTypes'
import type { Asset } from 'types/assets/asset'
import type { Valuations } from 'components/forms/ValuationsInput'

export type RealEstate = {
  name: string,
  address: Address,
  notes: string,
  purchaseDate: string,
  saleDate?: string,
  valuations: Valuations
}

export type RealEstateWithId = RealEstate & { id: string }

export type RealEstateMap = { [id: string]: RealEstate }

export const RealEstateDefaults = {
  name: '',
  address: AddressDefaults,
  notes: '',
  purchaseDate: '',
  valuations: [],
}

export const realEstateToAsset = (realEstate: RealEstate): Asset => {
  const asset = {}
  asset.type = 'RealEstate'
  asset.name = realEstate.name
  asset.startDate = realEstate.purchaseDate
  asset.endDate = realEstate.saleDate
  asset.valuations = realEstate.valuations
  if (typeof realEstate.id === 'number') { asset.instanceId = realEstate.id.toString() }
  return asset
}
