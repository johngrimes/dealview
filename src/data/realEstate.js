import { AddressDefaults } from './commonTypes.js'
import { saleNegFilter } from './valuations.js'

export const RealEstateDefaults = {
  name: '',
  address: AddressDefaults,
  notes: '',
  purchaseDate: '',
  valuations: [],
}

export const realEstateToAsset = realEstate => {
  const asset = {}
  asset.type = 'RealEstate'
  asset.name = realEstate.name
  asset.startDate = realEstate.purchaseDate
  asset.endDate = realEstate.saleDate
  // Sale valuations are omitted from the valuations that get stored against the asset.
  // The asset was never worth that much on our balance sheet - that value immediately went to cash, or whatever else.
  asset.valuations = realEstate.valuations.filter(saleNegFilter)
  if (typeof realEstate.id === 'string') {
    asset.id = realEstate.id
  }
  return asset
}
