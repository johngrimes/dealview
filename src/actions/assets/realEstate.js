// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import AssetActions from 'actions/assets/assets'
import { realEstateToAsset } from 'types/assets/realEstate'
import { createObjectActions } from 'actions/objects'
import type { RealEstate } from 'types/assets/realEstate'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset.RealEstate'

const actions = createObjectActions('REAL_ESTATE', 'REAL_ESTATE',
                                    'RealEstate', 'RealEstate')

const putRealEstate = (realEstate: RealEstate): Thunk => {
  return dispatch => {
    dispatch(actions.putRealEstateRequest(realEstate))
    return new Promise(async (resolve, reject) => {
      try {
        const saved = await putObject(objectStore, realEstate)
        dispatch(actions.putRealEstateSuccess(saved))
        await dispatch(AssetActions.putAsset(realEstateToAsset(saved)))
        resolve(saved)
      } catch (error) {
        dispatch(actions.putRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const deleteRealEstate = (id: string): Thunk => {
  return dispatch => {
    dispatch(actions.deleteRealEstateRequest(id))
    return new Promise(async (resolve, reject) => {
      try {
        const deletedId = await deleteObject(objectStore, id)
        dispatch(AssetActions.deleteAsset(deletedId))
        dispatch(actions.deleteRealEstateSuccess(deletedId))
        resolve(deletedId)
      } catch (error) {
        dispatch(actions.deleteRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const loadRealEstate = (): Thunk => {
  return dispatch => {
    dispatch(actions.loadRealEstateRequest())
    return new Promise((resolve, reject) => {
      getAllObjects(objectStore).then(assets => {
        dispatch(actions.loadRealEstateSuccess(assets))
        resolve(assets)
      }).catch(error => {
        dispatch(actions.loadRealEstateFailure(error))
        reject(error)
      })
    })
  }
}

export default {
  ...actions,
  putRealEstate,
  deleteRealEstate,
  loadRealEstate,
}
