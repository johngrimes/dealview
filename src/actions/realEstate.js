// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import { putAssetRequest, putAssetSuccess, putAssetFailure,
         deleteAssetRequest, deleteAssetSuccess, deleteAssetFailure } from 'actions/assets'
import { realEstateToAsset } from 'types/assets/realEstate'
import type { RealEstate, RealEstateWithId, RealEstateMap } from 'types/assets/realEstate'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset.RealEstate'
const assetObjectStore = 'Asset'

type PutRealEstateRequestAction = {
  type: 'PUT_REAL_ESTATE_REQUEST',
}
type PutRealEstateSuccessAction = {
  type: 'PUT_REAL_ESTATE_SUCCESS',
  realEstate: RealEstateWithId,
}
type PutRealEstateFailureAction = {
  type: 'PUT_REAL_ESTATE_FAILURE',
  error: string|null,
}

type DeleteRealEstateRequestAction = {
  type: 'DELETE_REAL_ESTATE_REQUEST',
}
type DeleteRealEstateSuccessAction = {
  type: 'DELETE_REAL_ESTATE_SUCCESS',
  id: string,
}
type DeleteRealEstateFailureAction = {
  type: 'DELETE_REAL_ESTATE_FAILURE',
  error: string|null,
}

type LoadRealEstateRequestAction = {
  type: 'LOAD_REAL_ESTATE_REQUEST'
}
type LoadRealEstateSuccessAction = {
  type: 'LOAD_REAL_ESTATE_SUCCESS',
  realEstate: RealEstateMap,
}
type LoadRealEstateFailureAction = {
  type: 'LOAD_REAL_ESTATE_FAILURE',
  error: string|null,
}

export type RealEstateAction = PutRealEstateRequestAction
                             | PutRealEstateSuccessAction
                             | PutRealEstateFailureAction
                             | DeleteRealEstateRequestAction
                             | DeleteRealEstateRequestAction
                             | DeleteRealEstateFailureAction
                             | LoadRealEstateRequestAction
                             | LoadRealEstateSuccessAction
                             | LoadRealEstateFailureAction

export const putRealEstateRequest = (): PutRealEstateRequestAction => {
  return { type: 'PUT_REAL_ESTATE_REQUEST' }
}

export const putRealEstateSuccess = (realEstate: RealEstateWithId): PutRealEstateSuccessAction => {
  return {
    type: 'PUT_REAL_ESTATE_SUCCESS',
    realEstate,
  }
}

export const putRealEstateFailure = (error: string|null): PutRealEstateFailureAction => {
  return {
    type: 'PUT_REAL_ESTATE_FAILURE',
    error,
  }
}

export const deleteRealEstateRequest = (): DeleteRealEstateRequestAction => {
  return { type: 'DELETE_REAL_ESTATE_REQUEST' }
}

export const deleteRealEstateSuccess = (id: string): DeleteRealEstateSuccessAction => {
  return {
    type: 'DELETE_REAL_ESTATE_SUCCESS',
    id,
  }
}

export const deleteRealEstateFailure = (error: string|null): DeleteRealEstateFailureAction => {
  return {
    type: 'DELETE_REAL_ESTATE_FAILURE',
    error,
  }
}

export const loadRealEstateRequest = (): LoadRealEstateRequestAction => {
  return { type: 'LOAD_REAL_ESTATE_REQUEST' }
}

export const loadRealEstateSuccess = (realEstate: RealEstateMap): LoadRealEstateSuccessAction => {
  return {
    type: 'LOAD_REAL_ESTATE_SUCCESS',
    realEstate,
  }
}

export const loadRealEstateFailure = (error: string|null): LoadRealEstateFailureAction => {
  return {
    type: 'LOAD_REAL_ESTATE_FAILURE',
    error,
  }
}

export const putRealEstate = (realEstate: RealEstate): Thunk => {
  return dispatch => {
    dispatch(putRealEstateRequest())
    return new Promise((resolve, reject) => {
      putObject(objectStore, realEstate).then(saved => {
        dispatch(putRealEstateSuccess(saved))
        const asset = realEstateToAsset(saved)
        dispatch(putAssetRequest())
        putObject(assetObjectStore, asset)
          .then(savedAsset => {
            dispatch(putAssetSuccess(savedAsset))
            resolve(saved)
          }).catch(error => {
            dispatch(putAssetFailure(error))
            dispatch(deleteRealEstateRequest())
            deleteObject(objectStore, saved.id).then(id => {
              dispatch(deleteRealEstateSuccess(id))
              reject(error)
            }).catch(error => {
              dispatch(deleteRealEstateFailure(error))
              reject(error)
            })
          })
      }).catch(error => {
        dispatch(putRealEstateFailure(error))
        reject(error)
      })
    })
  }
}

export const deleteRealEstate = (id: string): Thunk => {
  return dispatch => {
    dispatch(deleteRealEstateRequest())
    return new Promise((resolve, reject) => {
      deleteObject(objectStore, id).then(deletedId => {
        dispatch(deleteRealEstateSuccess(deletedId))
        dispatch(deleteAssetRequest())
        deleteObject(assetObjectStore, deletedId)
        .then(() => {
          dispatch(deleteAssetSuccess(deletedId))
          resolve(deletedId)
        }).catch(error => {
          dispatch(deleteAssetFailure(error))
          resolve(error)
        })
      }).catch(error => {
        dispatch(deleteRealEstateFailure(error))
        reject(error)
      })
    })
  }
}

export const loadRealEstate = (): Thunk => {
  return dispatch => {
    dispatch(loadRealEstateRequest())
    return new Promise((resolve, reject) => {
      getAllObjects(objectStore).then(assets => {
        dispatch(loadRealEstateSuccess(assets))
        resolve(assets)
      }).catch(error => {
        dispatch(loadRealEstateFailure(error))
        reject(error)
      })
    })
  }
}
