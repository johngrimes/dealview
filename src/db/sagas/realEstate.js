// @flow

import { takeEvery, call, put, all } from 'redux-saga/effects'

import * as RealEstateActions from 'actions/realEstate'
import { putObject, deleteObject, getAllObjects } from 'db/db'
import type { PutRealEstateRequestAction, DeleteRealEstateRequestAction } from 'actions/realEstate'

const objectStore = 'Asset.RealEstate'

export function* watchPutRealEstate() {
  yield takeEvery('PUT_REAL_ESTATE_REQUEST', putRealEstate)
}

export function* putRealEstate(action: PutRealEstateRequestAction) {
  try {
    const saved = yield call(putObject, objectStore, action.realEstate)
    yield put(RealEstateActions.putRealEstateSuccess(saved))
  } catch (error) {
    yield put(RealEstateActions.putRealEstateFailure(error))
  }
}

function* watchDeleteRealEstate() {
  yield takeEvery('DELETE_REAL_ESTATE_REQUEST', deleteRealEstate)
}

function* deleteRealEstate(action: DeleteRealEstateRequestAction) {
  try {
    const id = yield call(deleteObject, objectStore, action.id)
    yield put(RealEstateActions.deleteRealEstateSuccess(id))
  } catch (error) {
    yield put(RealEstateActions.deleteRealEstateFailure(error))
  }
}

function* watchLoadRealEstate() {
  yield takeEvery('LOAD_REAL_ESTATE_REQUEST', loadRealEstate)
}

function* loadRealEstate() {
  try {
    const realEstate = yield call(getAllObjects, objectStore)
    yield put(RealEstateActions.loadRealEstateSuccess(realEstate))
  } catch (error) {
    yield put(RealEstateActions.loadRealEstateFailure(error))
  }
}

export default function* watchRealEstate(): void {
  yield all([
    watchPutRealEstate(),
    watchDeleteRealEstate(),
    watchLoadRealEstate(),
  ])
}
