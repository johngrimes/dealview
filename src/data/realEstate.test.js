import { realEstateToAsset } from './realEstate.js'
import { validRealEstate2 } from './fixtures/realEstate'
import { saleFilter } from './valuations.js'

describe('realEstateToAsset', () => {
  it('should convert a real estate asset with a sale correctly', () => {
    const realEstate = { ...validRealEstate2 }
    expect(realEstateToAsset(realEstate)).toMatchSnapshot()
  })

  it('should omit the sale price in the asset valuations', () => {
    const realEstate = { ...validRealEstate2 }
    const asset = realEstateToAsset(realEstate)
    expect(asset.valuations.filter(saleFilter)).toHaveLength(0)
  })
})
