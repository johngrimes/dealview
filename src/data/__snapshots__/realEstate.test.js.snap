// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`realEstateToAsset should convert a real estate asset with a sale correctly 1`] = `
Object {
  "endDate": "2017-01-01",
  "name": "38/8 Briggs Rd",
  "startDate": "2005-05-10",
  "type": "RealEstate",
  "valuations": Array [
    Object {
      "amount": 165000,
      "date": "2005-05-10",
      "note": "Purchase price",
      "type": "purchase",
    },
    Object {
      "amount": 260000,
      "date": "2015-09-10",
      "note": "Formal valuation",
      "type": "none",
    },
  ],
}
`;
